Add support for NSSL 2-moment cloud microphysics scheme

MicroTed opened the pull request:
NSSL 2-moment Cloud Microphysics (National Severe Storms Laboratory)

Features consistent with WRF-ARW release (same code base):

- Fully 2-moment (mass and number), for droplets, rain, ice crystal, snow, graupel, and hail (hail can be deactivated)
- Predicted bulk concentration of cloud condensation nuclei (CCN) (optional)
- Allows residual supersaturation (no saturation adjustment) for reasonably physical aerosol response
- Adaptive 2-moment sedimentation to prevent excessive size sorting
- Predicts bulk density of graupel and hail
- Hail species intended as “true” hail, characterized by size and source rather than simple density
- Designed primarily for convection-allowing grid spacing (dx <= ~4km)
- Computes effective radii for radiation (sets effr_in=.true.)
- Computes radar reflectivity assuming Rayleigh scattering (sets ltradar=.true.)

Main reference: Mansell, E. R., C. L. Ziegler, and E. C. Bruning, 2010: Simulated electrification of a small thunderstorm with two-moment bulk microphysics. J. Atmos. Sci., 67, 171-194, doi:10.1175/2009JAS2965.1

Namelist options:

nssl_cccn - CCCN background CCN concentration. Default: 0.6e9 (per m^3 at STP)
nssl_alphah - graupel shape parameter, default = 0.
nssl_alphahl - hail shape parameter, default = 1.
nssl_hail_on - hail activation flag, default = .false.
nssl_ccn_on - CCN prediction flag, default = .true.
nssl_invertccn - store CCN as number of activated instead of number remaining, default = .true. (avoids need for specified boundary conditions)

Tested with SCM (pull request initiated) and earlier version tested with UFS-LAM
e.g.,
./run_scm.py -c twpice -n input_GSD_v1nssl_nohailnoccn.nml -t tracers_gsd_nssl_nohail_noccn.txt -s SCM_GSD_v1nssl

Related PRs:
For SCM: https://github.com/NCAR/ccpp-scm/pull/277
For FV3: https://github.com/NOAA-EMC/fv3atm/pull/472

MicroTed commented:
FYI @tsupinie

MicroTed commented:
Companion SCM pull request is https://github.com/NCAR/ccpp-scm/pull/277

grantfirl commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -77,14 +78,17 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
                                            make_DropletNumber,       &
                                            make_RainNumber
 
+!      use module_mp_nssl_2mom,       only: calc_eff_radius, calcnfromq, na
```
Please remove this comment.

grantfirl commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -666,6 +684,33 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
+              enddo
+            enddo
+            ENDIF
+!            write(6,*) 'rrtmg_pre: max qctrac,qc,qcphy,nctrac,ccw,ccwphy: ',maxval(qc_mp),maxval(qc), &
```
Please remove debugging comments if possible.

grantfirl commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -757,7 +802,116 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
+             enddo
+           enddo
+          else
+#if 0
```
What is the story with this section? Do you still need it? If so, could it fit within a timestep_init phase rather than the run phase?

MicroTed replied:
Sorry, I missed cleaning that up. I was trying to make the calls for calculating radii from the rrtmg subroutine, but was getting blocks that randomly had no number concentrations (with fv3atm, but code as of early 2021). Maybe with current fv3atm it should work, but it is not necessary so I'll take it out (along with the "use" statement that it needed). Initializing arrays from mp_nssl_init also was not working with that code, but it seems to work in SCM.

climbfuji commented:
> @grantfirl I haven't made a PR for fv3atm (develop) or atmos_cubed_sphere (dev/emc), but I have updated to recent code at
> 
> https://github.com/MicroTed/fv3atm.git feature/nsslmicro
> 
> https://github.com/MicroTed/GFDL_atmos_cubed_sphere.git feature/nssl-base
> 
> The reason being that although I've been able to compile those, I haven't gotten it to run yet. It completely fails off the bat on Mac, and my guess is something with ESMF. And on Jet it gets through the init phase and then mysteriously dies (LAM setup)

Are you using clang + gfortran? There is a known issue since ESMF 8.1.1 (I think) that the code crashes immediately. It works with GNU gcc+gfortran.

grantfirl commented on physics/mp_nssl.F90:
```diff
@@ -0,0 +1,962 @@
+                              
+
+        use module_mp_nssl_2mom, only: nssl_2mom_init, calcnfromq, na
+        use physcons, only: con_rd
```
Using the physcons module is going to be a problem. To the extent that other CCPP physics are doing this, it should be getting cleaned up. The reason is that the physcons module will not be part of the CCPP physics repo longterm, and instead, physical constants must be defined by (and have metadata) in the host model. The result is that physics schemes that need physical constants must get them from the argument list. Please see https://ccpp-techdoc.readthedocs.io/en/latest/CompliantPhysicsParams.html#using-constants for more info.

MicroTed commented on physics/GFS_suite_interstitial.F90:
```diff
@@ -576,9 +581,10 @@ subroutine GFS_suite_interstitial_3_run (im, levs, nn, cscnv,       &
+!          if ( n /= ntcw  .and. n /= ntiw  .and. n /= ntclamt .and. &
+!               n /= ntrw  .and. n /= ntsw  .and. n /= ntrnc   .and. &
+!               n /= ntsnc .and. n /= ntgl  .and. n /= ntgnc) then
+            IF ( otsptflag(n) ) THEN
```
Yeah, I wasn't sure if that should be separate, as I also don't know what the rationale for excluding variables is, and was guessing based on what was already there. Mainly I was trying to avoid adding more and more indices to the call interface. Either way is fine with me.

grantfirl commented on physics/mp_nssl.F90:
```diff
@@ -0,0 +1,962 @@
+         
+         ENDIF ! .not. is_initialized
+         
+#if 0
```
This section may fall in the same category as the other #if 0 section? If so, please remove.

MicroTed replied:
I've done some testing with FV3 (ufs_weather_model), and for some reason the init routine has no effect on the input tracers arrays. I can set values for zero number concentrations (values not in the IC data), but they are zero again when it gets to the nssl_run routine. The cap routines have GFS_Data as inout, so that shouldn't be an issue. (it is not just the number concentrations -- I did a test (although with older code) where I set all the mixing ratios to zero in the init routine, and that also had no effect.)

grantfirl commented on physics/mp_nssl.F90:
```diff
@@ -0,0 +1,962 @@
+        ENDIF
+       
+       IF ( .true. ) THEN
+        DO n = 1,ntmul
```
OK, I suppose the previous comment can be disregarded. I see that you've implemented an internal loop here.

MicroTed replied:
Yeah, that originated with my first integration with SCM, which uses large time steps like 600s. The NSSL scheme isn't designed for that, and some sensitivity testing resulted in about a 2-3min time step working OK.

MicroTed commented:
> Hmm no, that should work. I am using gcc+gfortran 10 with ESMF 8.2.0 bs14 and that works fine. Did you reduce the compiler optimization? I remember this was something that was required on macOS. If you build ESMF with hpc-stack, then all this is taken care of.

I actually wasn't aware of hpc-stack. I can give that a try. I've been building from the NCEPlibs and NCEPlibs-external.

climbfuji commented:
> > Hmm no, that should work. I am using gcc+gfortran 10 with ESMF 8.2.0 bs14 and that works fine. Did you reduce the compiler optimization? I remember this was something that was required on macOS. If you build ESMF with hpc-stack, then all this is taken care of.
> 
> I actually wasn't aware of hpc-stack. I can give that a try. I've been building from the NCEPlibs and NCEPlibs-external.

Hmm. Those should be ok, if I remember correctly. Pease make sure that ESMF is compiled without optimization, for example:
```
# ESMF_BOPT: O
# ESMF_OPTLEVEL: 0
```

climbfuji commented:
> > > Hmm no, that should work. I am using gcc+gfortran 10 with ESMF 8.2.0 bs14 and that works fine. Did you reduce the compiler optimization? I remember this was something that was required on macOS. If you build ESMF with hpc-stack, then all this is taken care of.
> > 
> > 
> > I actually wasn't aware of hpc-stack. I can give that a try. I've been building from the NCEPlibs and NCEPlibs-external.
> 
> Hmm. Those should be ok, if I remember correctly. Pease make sure that ESMF is compiled without optimization, for example:
> 
> ```
> # ESMF_BOPT: O
> # ESMF_OPTLEVEL: 0
> ```

Check also here for Darwin-specific settings:

https://github.com/NOAA-EMC/hpc-stack/blob/develop/libs/build_esmf.sh

MicroTed commented:
> Apparently, GitHub doesn't want to show us all 20K lines of the module_mp_nssl_2mom.F90 file. Physcons is used in there and should also be changed for the same reasons as mentioned previously. I cannot pretend to have carefully reviewed all lines of this file, but I did notice a few things that a scheme should not have to be truly CCPP-compliant. First, the scheme should not be allowed to STOP the model. One should pass in the errflag and errmsg variables and set them as appropriate and return rather than STOP. I also noticed a wrf_error_fatal call in there (although it looked like the abort command was commented out). I also noticed a lot of debugging WRITE calls. Could these be wrapped in a debug flag or something?

I'm working the physcon issue, which is not a big deal (constants were only used in the top module section). I had alternate subroutines for the "wrf_xxx" calls, but there are so few that I'll switch those to use errmsg/errflg if possible.

MicroTed commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -1108,5 +1196,4 @@ end subroutine GFS_rrtmg_pre_run
       subroutine GFS_rrtmg_pre_finalize ()
       end subroutine GFS_rrtmg_pre_finalize
 
-!! @}
```
Hmm, I don't recall touching that, but I'll put that back in.

MicroTed commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -171,6 +176,7 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
 
       character(len=*), intent(out) :: errmsg
       integer,          intent(out) :: errflg
+      integer,          intent(in)  :: mpiroot
```
Will remove

MicroTed commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -213,6 +223,7 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
                  its, ite, jts, jte, kts, kte
 
       real(kind=kind_phys) :: qvs
+      real (kind=kind_phys) :: sum1,sum2,max1,max2
```
Not any more... will remove

MicroTed commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -191,6 +197,10 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
+      ! for NSSL MP
+      real(kind=kind_phys), dimension(im,lm+LTP) ::         &
+                             re_rain, qr_mp, ns_mp, nr_mp, nh_mp, vh_mp, cccn_mp,cccna_mp, nc_mp2
+      real, allocatable :: an(:,:,:,:) ! temporary scalar array
```
Code was removed -- so "an" can go. The others are not needed at this time, either, so they can go, too. (I was attempting to calculate radii within the subroutine, but it has a weird problem in FV3 that I couldn't isolate.)

MicroTed commented on physics/GFS_suite_interstitial.F90:
```diff
@@ -529,9 +531,12 @@ subroutine GFS_suite_interstitial_3_run (im, levs, nn, cscnv,       &
       implicit none
 
       ! interface variables
+      logical, intent(in)     :: otsptflag(1:ntracp1)!  on/off switch for tracer transport
```
Right. And 'ntrapcp1' is not needed -- otsptflag only needs to be the size of 'ntrac'

MicroTed commented:
@grantfirl -- Arguments have all been set to "optional = F" 

> @MicroTed According to [NCAR/ccpp-framework#408](https://github.com/NCAR/ccpp-framework/pull/408), the optional arguments are no longer going to be supported in the CCPP.

grantfirl commented:
> @grantfirl -- Arguments have all been set to "optional = F" -- or do these lines need to be removed altogether?
> 
> > @MicroTed According to [NCAR/ccpp-framework#408](https://github.com/NCAR/ccpp-framework/pull/408), the optional arguments are no longer going to be supported in the CCPP.

Yes, please remove all optional attributes in your .meta files. Would you like me to help update this PR to the latest main?

MicroTed commented:
@grantfirl 
I rebased to the current ccpp/main and removed those lines. Right now having some weird problem pushing the update.

> Yes, please remove all optional attributes in your .meta files. Would you like me to help update this PR to the latest main?

grantfirl commented:
> @grantfirl I rebased to the current ccpp/main and removed those lines. Right now having some weird problem pushing the update.
> 
> Update: I think that succeeded. SCM doesn't build, though. Maybe the CCPP switch from "various" to "mixed" in the .meta files?
> 
> > Yes, please remove all optional attributes in your .meta files. Would you like me to help update this PR to the latest main?

Yes, SCM main is behind ccpp-physics a few PRs. Will fix soon.

MicroTed commented:
> Yes, SCM main is behind ccpp-physics a few PRs. Will fix soon.

Aside: Are there models besides SCM and UFS/FV3 that are using CCPP and worth a little time to check for adding compatibility?

climbfuji commented:
> > Yes, SCM main is behind ccpp-physics a few PRs. Will fix soon.
> 
> Aside: Are there models besides SCM and UFS/FV3 that are using CCPP and worth a little time to check for adding compatibility?

NEPTUNE. But we can't see the code. Maybe you can. You would need to check with @matusmartini if they are interested and willing to help testing. But their code is usually behind and it may be difficult to test your new scheme.

MicroTed commented:
Updated to recent CCPP (~28 Dec. 2021) that works with latest SCM.

MicroTed commented on physics/GFS_suite_interstitial.F90:
```diff
@@ -698,23 +713,29 @@ end subroutine GFS_suite_interstitial_4_finalize
 
       ! interface variables
 
+      logical, intent(in)     :: otsptflag(1:ntracp1)! on/off switch for tracer transport by updraft and
```
Fixed

MicroTed commented:
Also updated SCM PR with current CCPP
https://github.com/NCAR/ccpp-scm/pull/277/files

I assume I should add some scheme documentation in physics/physics/docs?

grantfirl commented:
@climbfuji Since you requested changes on this PR, you will need to review again.

climbfuji commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -666,6 +682,26 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
+                ni_mp (i,k) = tracer1(i,k,ntinc)/(1.-qvs)
+                ns_mp (i,k) = tracer1(i,k,ntsnc)/(1.-qvs)
+                nr_mp (i,k) = tracer1(i,k,ntrnc)/(1.-qvs)
+                IF ( nssl_ccn_on ) cccn_mp(i,k) = tracer1(i,k,ntccn)/(1.-qvs)
```
Same as above - move the if block outside the nested do loops. And possible define a local variable `1/(1.-qvs)` for better computational performance

climbfuji replied:
But in this case it would be good to define a constant `1./(1.-qvs) = 1./(1.-tracer1(i,k,ntqv))` and use that.

grantfirl replied:
I'm so confused by this, but this comment is attached to an older version of this code or something. This block of changes doesn't exist in the latest code.

climbfuji commented on physics/module_MYNNPBL_wrapper.F90:
```diff
@@ -843,6 +880,21 @@ SUBROUTINE mynnedmf_wrapper_run(        &
+                 dqdt_cloud_droplet_num_conc(i,k)  = RQNCBLTEN(i,k)
+                 dqdt_ice_cloud(i,k)               = RQIBLTEN(i,k) !/(1.0 + qv(i,k))
+                 dqdt_ice_num_conc(i,k)            = RQNIBLTEN(i,k)
+                 IF ( nssl_ccn_on ) THEN ! 
```
Same as above

grantfirl commented:
@MicroTed I'll make the changes that @climbfuji requested, retest on Hera/intel, and submit a PR into your branch.

MicroTed commented on physics/GFS_PBL_generic.F90:
```diff
@@ -546,6 +617,57 @@ subroutine GFS_PBL_generic_post_run (im, levs, nvdiff, ntrac,
+               dqdt(i,k,ntgv) = dvdftra(i,k,14) 
+               dqdt(i,k,nthv) = dvdftra(i,k,15) 
+               dqdt(i,k,ntoz) = dvdftra(i,k,16) 
+               IF ( nssl_ccn_on ) THEN
```
That’s a good point, but don’t modern compilers usually take care of this as part of optimization? Basically the compiler should make two full loops, one with ccn and the other without, so that only one loop runs. A separate loop for one variable might go slightly  slower?

climbfuji replied:
Possible. And yes, you are right, the compiler does all sorts of things. It might even do the same with my suggested approach of having two loops. Please ignore my comments regarding those optimizations.

grantfirl replied:
I had implemented this for GFS_PBL_generic.F90. Should I not commit this now?

climbfuji replied:
Sorry. Ted made a good point that I wasn't thinking of.

grantfirl replied:
So, @climbfuji, does this mean that you approve of the code as-is?

climbfuji commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -666,6 +682,26 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
+            IF ( .not. effr_in ) THEN
+            do k=1,LMK
+              do i=1,IM
+                qvs = qgrs(i,k,ntqv)
```
@grantfirl I am still seeing this code ... but maybe github is playing tricks with me?

grantfirl replied:
@climbfuji If I click on the "View changes" button next to your comment, it takes me to this file that shows changes through 25 commits. If instead, starting from the main pull request and clicking on "Files Changed", it shows changes from through all commits, and this change is no longer there. I don't understand why you're seeing code through 25 commits instead of 71.

MicroTed replied:
That is old code that should be gone. It was an attempt to calculate effective radii locally when needed instead of in the microphysics for every time step. I may try again at some point, but I was getting really weird behavior where some input array values randomly were all zero. So I gave up for the time being.

climbfuji replied:
Ok ok ... not my github day today it seems. All good and sorry for the confusion.

MicroTed commented:
Some internal documentation has been added (comments for subroutines, etc. for doxygen). Plus added file in docs/pdftxt for the NSSL scheme. Thanks to Man Zhang for help!

grantfirl commented:
@MicroTed This PR is the next one in the queue. Please pull down the latest main, merge it into this branch, and push back to GitHub. If there are merge conflicts (which it appears there may be since this PR and main have both modified a couple of files), it looks like you just need to accept the union of the two changes, and make sure that there is no duplication.

MicroTed commented:
@grantfirl  It is updated and a test run of FV3 was successful.

grantfirl commented:
@MicroTed We need to update this pull request with #869 in order to fix a compilation memory error on wcoss_dell that is largely unrelated to anything having to do with NSSL MP. Here are the steps that I would suggest:

1. Starting from within ccpp/physics on your local machine, let's add Dom Heinzeller's fork as a remote (since that is where #869 is coming from): `git remote add dom-fork https://github.com/climbfuji/ccpp-physics`
2. Fetch Dom's branches: `git fetch dom-fork`
3. Checkout #869 PR branch: `git checkout feature/reduce_opt_gfs_typedefs`
4. Merge this into your feature/nsslmicro branch: `git checkout feature/nsslmicro`, `git merge feature/reduce_opt_gfs_typedefs`
5. Push to Github

Once this is done, you'll need to update your FV3 PR branch as well:

1. cd up to FV3
2. `git add ccpp/physics`, `git commit -m "update ccpp/physics submodule pointer"`
3. Push to GitHub

Once that is done, I'll update the ufs-weather-model PR branch.

MicroTed commented:
@grantfirl OK, ccpp/physics and fv3atm have been pushed.
